### PR TITLE
Issus 4908 port save file dialog designer and folder browser dialog designer

### LINKED ISSUE #4908: Support designer capabilities at runtime
# Context

The Windows Forms SDK consist of two distinct parts – the **runtime** (the code that executes, open sourced in this repo) and the **designer** (consists of the "_general purpose_" and "_Visual Studio-specific_"  components, close sourced). In .NET Framework these two parts lived together, and that allowed customers to invoke and use VS-specific functionality. Skip forward to .NET Core/.NET and these two components are now split and evolve (for the most part) independently of each other.
 
The more and more customers are starting to migrate their apps to .NET the more we see the demand to the missing designer capabilities to be ported over to .NET (e.g. #1395, #2573, #4456, #4886, #4887, [AB#1314797](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1314797)).

# In scope:

However we need to port the general purpose designer infrastructure – i.e. API that enable building "_a designer that the user can embed in their application_". For example, a user might have a business application that contains a report designer feature, which can use our "general designer" framework.

* The following designers are currently available (though may have restricted or incomplete functionality):
  - [x] System.ComponentModel.Design.ComponentDesigner
  - [x] System.Windows.Forms.Design.DocumentDesigner (#122)
  - [x] System.Windows.Forms.Design.FormDocumentDesigner (#681)
  - [x] System.Windows.Forms.Design.MaskedTextBoxDesigner (#9709)
  - [x] System.Windows.Forms.Design.ParentControlDesigner (#122)
  - [x] System.Windows.Forms.Design.ScrollableControlDesigner (#122)
  - [x] System.Windows.Forms.Design.TextBoxBaseDesigner
  - [x] System.Windows.Forms.Design.ToolStripDesigner
  - [x] System.Windows.Forms.Design.ToolStripDropDownDesigner
  - [x] System.Windows.Forms.Design.ToolStripItemDesigner
  - [x] System.Windows.Forms.Design.ToolStripMenuItemDesigner
 
* The following designers are missing and will be ported in scope of #4860:
  - [x] System.Windows.Forms.Design.ButtonBaseDesigner
  - [x] System.Windows.Forms.Design.ControlDesigner
  - [x] System.Windows.Forms.Design.DocumentDesigner - functionality restored
  - [x] System.Windows.Forms.Design.LabelDesigner
  - [x] System.Windows.Forms.Design.ParentControlDesigner - functionality restored
  - [x] System.Windows.Forms.Design.ScrollableControlDesigner - functionality restored
  - [x] System.Windows.Forms.Design.TextBoxDesigner
  - [x] System.Windows.Forms.Design.ComboBoxDesigner
  - [x] System.Windows.Forms.Design.GroupBoxDesigner
  - [x] System.Windows.Forms.Design.ListBoxDesigner
  - [x] System.Windows.Forms.Design.ListViewDesigner
  - [x] System.Windows.Forms.Design.PanelDesigner
  - [x] System.Windows.Forms.Design.PictureBoxDesigner
  - [x] System.Windows.Forms.Design.RadioButtonDesigner
  - [x] System.Windows.Forms.Design.RichTextBoxDesigner
  - [x] System.Windows.Forms.Design.TreeViewDesigner
  - [x] System.Windows.Forms.Design.UpDownBaseDesigner
  - [x] System.Windows.Forms.Design.UserControlDocumentDesigner

# Open for consideration

The following API may be considered for porting upon significant demand and real use cases.

* The following designers:
  - [x] System.Windows.Forms.Design.DateTimePickerDesigner (#8549)
  - [x] System.Windows.Forms.Design.FlowLayoutPanelDesigner (#9658 9658)
  - [x] System.Windows.Forms.Design.MonthCalendarDesigner (#9536)
  - [x] System.Windows.Forms.Design.PrintDialogDesigner (#9595  )
  - [x] System.Windows.Forms.Design.TableLayoutPanelDesigner(#9560)
  - [x] System.Windows.Forms.Design.TabControlDesigner(#9560)
  - [x] System.Windows.Forms.Design.TabPageDesigner(#9560)
* Missing editors for the ported designers, see the list in #2411

# Out of scope

* Back-porting of the missing types in .NET Core 3.1 or .NET 5.0, because both of these are going out of support shortly after .NET 6.0 is released.
* The VS-specific designer functionality will not be made available, because it makes no sense outside the VS. 
* Databinding
* Advanced databinding designers and editors
* The following designers:
  - [x] System.Windows.Forms.Design.AxDesigner (Tracked with #9810)
  - [x] System.Windows.Forms.Design.AxHostDesigner  (Tracked with #9810)
  - [x] System.Windows.Forms.Design.BindingNavigatorDesigner(https://github.com/dotnet/winforms/pull/9749)
  - [x] System.Windows.Forms.Design.BindingSourceDesigner(https://github.com/dotnet/winforms/pull/9749)
  - [x] System.Windows.Forms.Design.DataGridViewColumnDesigner (https://github.com/dotnet/winforms/pull/9677)
  - [x] System.Windows.Forms.Design.DataGridViewComboBoxColumnDesigner (https://github.com/dotnet/winforms/pull/9869)
  - [x] System.Windows.Forms.Design.DataGridViewDesigner (https://github.com/dotnet/winforms/pull/9677)
  - [x] System.Windows.Forms.Design.FolderBrowserDialogDesigner (https://github.com/dotnet/winforms/pull/9682)
  - [x] System.Windows.Forms.Design.ImageListDesigner
  - [x] System.Windows.Forms.Design.NotifyIconDesigner (https://github.com/dotnet/winforms/pull/9645)
  - [x] System.Windows.Forms.Design.PropertyGridDesigner (https://github.com/dotnet/winforms/pull/9645)
  - [x] System.Windows.Forms.Design.SaveFileDialogDesigner (https://github.com/dotnet/winforms/pull/9682)
   <del>System.Windows.Forms.Design.StatusBarDesigner</del>
  - [x] System.Windows.Forms.Design.SplitContainerDesigner (https://github.com/dotnet/winforms/pull/9320)
  - [x] System.Windows.Forms.Design.SplitterDesigner (https://github.com/dotnet/winforms/pull/9320)
  - [x] System.Windows.Forms.Design.SplitterPanelDesigner (https://github.com/dotnet/winforms/pull/9320)
  - [x] System.Windows.Forms.Design.ToolStripContainerDesigner (https://github.com/dotnet/winforms/pull/9750)
  - [x] System.Windows.Forms.Design.ToolStripContentPanelDesigner (https://github.com/dotnet/winforms/pull/9750)
  - [x] System.Windows.Forms.Design.ToolStripPanelDesigner (https://github.com/dotnet/winforms/pull/9750)
  - [x] System.Windows.Forms.Design.TrackBarDesigner (https://github.com/dotnet/winforms/pull/9703)
  - [x] System.Windows.Forms.Design.WebBrowserDesigner (Tracked with #9810)

### PR DESCRIPTION
**Pull request by LeafShi1**
Fixes #4908 

## Proposed changes

- Port SaveFileDialogDesigner and FolderBrowserDialogDesigner to runtime

## Customer Impact

- SaveFileDialogDesigner and FolderBrowserDialogDesigner can be designed in runtime

## Regression? 

- No

## Risk

- Minimal

### Before

BeforeChang

### After

AfterChang

## Test methodology 

- Manually (added SaveFileDialogDesigner and FolderBrowserDialogDesigner to DemoConsole test app)

## Test environment(s) 

- .net 8.0.100-rc.1.23407.2

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9682)

### CODE REVIEW

**Review by SimonZhao888 — DISMISSED** [review]

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs [hunk 1]
```diff
@@ -34,6 +34,23 @@ public partial class ComponentDesigner : ITreeDesigner, IDesignerFilter, ICompon
     /// </summary>
     public virtual ICollection AssociatedComponents => Array.Empty<IComponent>();
 
+    private protected virtual void UpdateTextualDefaultProperty()
+    {
+        var component = Component;
+        if (component?.Site is { } site)
+        {
+            var defaultProperty = TypeDescriptor.GetDefaultProperty(component);
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs:42]: We declare type explicitly if it isn't obvious from reading code.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs [hunk 2]
```diff
@@ -34,6 +34,23 @@ public partial class ComponentDesigner : ITreeDesigner, IDesignerFilter, ICompon
...
+            if (defaultProperty is not null && defaultProperty.PropertyType.Equals(typeof(string)))
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs:43]: nit: reduce nesting by reverting `if` clause.

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs [hunk 3]
```diff
@@ -125,6 +142,15 @@ public virtual void InitializeNewComponent(IDictionary defaultValues)
     {
         // execute legacy code
         InitializeNonDefault();
+
+        // Note: This was originally an obsoleted API called OnSetComponentDefaults(). The
+        // default behavior of this API was to set the the default property to the component's
+        // site name, if the property was a string and null or empty. We've removed the API
+        // but preserved the same behavior, now controlled by SetTextualDefaultProperty.
+        if (SetTextualDefaultProperty)
```
Comment (dreddy-work) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs:150]: this is now applied for all component designers but not just the ones being ported? was there a bug for other component designers prior to this change?

Comment (LeafShi1) [reply src/System.Windows.Forms.Design/src/System/ComponentModel/Design/ComponentDesigner.cs:150]: Yes, this now applies to all component designers.
Before adding this judgment, there are 3 controls whose default properties are different from the Winform designer, other controls are not affected by this.

**The 3 controls -> property:**
image

File: src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/FolderNameEditor.FolderBrowser.cs [hunk 4]
```diff
@@ -94,7 +94,7 @@ public unsafe DialogResult ShowDialog(IWin32Window? owner)
                     }
 
                     // Retrieve the path from the IDList.
-                    PWSTR selectedPath = default;
+                    PWSTR selectedPath = pDisplayName;
```
[the remark this answers is not in the document]
Comment (LeafShi1) [reply src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/FolderNameEditor.FolderBrowser.cs:97]: I don't know why the value was set to "default" before, but the default value always points to "0", and when I set the property value, such as FolderBrowerDialog's SelectedPath, an exception occurs.
image


Compare with the code with [NexFx](https://devdiv.visualstudio.com/DevDiv/_git/NetFx?path=/src/NDP/fx/src/Designer/WinForms/System/WinForms/Design/FolderNameEditor.cs&version=GBmain&line=229&lineEnd=229&lineStartColumn=21&lineEndColumn=86&lineStyle=plain&_a=contents), the selectedPath should use pDisplayName pointer.

**Review by dreddy-work — APPROVED** [review]

**Comment by dreddy-work** [comment]
please rebase.

**Comment by Tanya-Solyanik** [comment]
Looks like `feature/9.0` branch took a mad merge - nullability is missing in public API file?

src\System.Windows.Forms\src\System\Windows\Forms\TreeNodeCollection.cs(212,25): error RS0036: (NETCORE_ENGINEERING_TELEMETRY=Build) Symbol 'virtual System.Windows.Forms.TreeNodeCollection.AddRange(params System.Windows.Forms.TreeNode![]! nodes) -> void' is missing nullability annotations in the declared API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)

**Comment by LeafShi1** [comment]
> nullability

So we need to add` #nullable enable`  in [PublicAPI.Unshipped.txt](https://github.com/dotnet/winforms/blob/feature/9.0/src/System.Windows.Forms.Design/src/PublicAPI.Unshipped.txt)

**Review by Epica3055 — APPROVED** [review]

**Comment by Ashley-Li** [comment]
Verified this PR in feature/9.0 branch, now SaveFileDialogDesigner and FolderBrowserDialogDesigner are supported in runtime, we can modify them in DemoConsole app by its some properties:

https://github.com/dotnet/winforms/assets/56664234/5612ae30-12fb-4dc8-a72d-858b6e8d25dc

**Pull request merged**